### PR TITLE
Use-case article: Knowledge Graph Embedding

### PR DESCRIPTION
**Pull request by ricsi98**
This article is about how we can use KGE for simple question answering and how it might be better than using LLMs when we are provided with a KG as data source.

Outline:
- What are KGs?: intro to KGs
- What is KGE?: intro to KGE with a concrete example (DistMult algorithm)
- Build and Train a KGE model: example code with explanation using the DistMult model on the FreeBase15k dataset
- Answering questions with the model: code example with explanation how the previously trained model can be used for QA with a concrete question
- Comparing KGE with LLM performance on a large KG: Results of an experiment on a medium sized dataset that support the claim: KGE algos are better than LLMs for QA when the data is in a KG format.
- Limitations

### CODE REVIEW

File: docs/use_cases/knowledge_graph_embedding.md [hunk 1]
```diff
@@ -0,0 +1,228 @@
+<!-- SEO: Introduction to Knowledge Graphs. Introduction to Knowledge Graph Embedding. Introduction to the DistMult algorithm. Example code for training KGE model. Example code for evaluating KGE model. Example code for answering questions with a KGE model. KGE and LLM comparison on QA task.  -->
+
+# Learning Scemantic Representations for Knowledge Graphs
+
+<!-- TODO: Cover image:
```
Comment (morkapronczay) [inline docs/use_cases/knowledge_graph_embedding.md:5]: Should this be here?

File: docs/use_cases/knowledge_graph_embedding.md [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+
+# Learning Scemantic Representations for Knowledge Graphs
+
+<!-- TODO: Cover image: 
+1. You can create your own cover image and put it in the correct asset directory,
+2. or you can give an explanation on how it should be and we will help you create one. Please tag arunesh@superlinked.com or @AruneshSingh (GitHub) in this case. -->
+
+In today's world, LLMs (Large Language Models) are everywhere, doing all sorts of language-related tasks really well. They're the go-to solution for understanding and using text in many different ways.
+
+However, in some specific areas, there are other more specialized approaches that might give better results for specific tasks.
```
Comment (morkapronczay) [inline docs/use_cases/knowledge_graph_embedding.md:11]: I'd ditch one of the specifics. Also, this could be part of the previous paragraph IMHO (1 sentence paragraphs are weird)

Comment (morkapronczay) [reply docs/use_cases/knowledge_graph_embedding.md:11]: for example I'd use particular in one of the instances

File: docs/use_cases/knowledge_graph_embedding.md [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+
+<!-- TODO: Cover image: 
+1. You can create your own cover image and put it in the correct asset directory,
+2. or you can give an explanation on how it should be and we will help you create one. Please tag arunesh@superlinked.com or @AruneshSingh (GitHub) in this case. -->
+
+In today's world, LLMs (Large Language Models) are everywhere, doing all sorts of language-related tasks really well. They're the go-to solution for understanding and using text in many different ways.
+
+However, in some specific areas, there are other more specialized approaches that might give better results for specific tasks. 
+
+This post digs into when these popular language models might not be the best choice. We'll see how specialized methods, like KGE (Knowledge Graph Embedding) algorithms, could actually be better for certain tasks.
```
Comment (morkapronczay) [inline docs/use_cases/knowledge_graph_embedding.md:13]: I'd hint here, that LLM's are great for text, but there are other type of data representations where, maybe not so much

File: docs/use_cases/knowledge_graph_embedding.md [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+
+“… in Freebase, 93.8% of people have no place of birth and [78.5% have no nationality](https://aclanthology.org/P09-1113.pdf), about 68% of people [do not have any profession](https://dl.acm.org/doi/abs/10.1145/2566486.2568032), while in Wikidata, about [50% of artists have no date of birth](https://arxiv.org/abs/2207.00143), and only [0.4% of known buildings have information about height](https://dl.acm.org/doi/abs/10.1145/3485447.3511932).”
+
+These imperfections, whether minor or major, can pose significant difficulties if we solely rely on the graph for information. In such a scenario, KGE algorithms prove to be extremely beneficial. They allow us to retrieve the missing data by utilizing the learned semantic meaning of nodes and links.
```
Comment (morkapronczay) [inline docs/use_cases/knowledge_graph_embedding.md:31]: I don't get this, and I consider myself only median stupid :)
What can we do exactly? Have a guess on the missing data?

File: docs/use_cases/knowledge_graph_embedding.md [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+
+### How KGE algorithms work?
+
+In general, KGE algorithms work by defining a similarity function in the embedding space, and learn the model by minimizing a loss function that penalizes the discrepancy between this similarity function and some notion of similarity in the graph. KGE algorithms can differ in the choice of the similarity function and the definition of node similarity in the graph.
```
Comment (morkapronczay) [inline docs/use_cases/knowledge_graph_embedding.md:35]: I'd do the first long sentence in 2 sentences:
1. Similarity in the graph
2. Loss function that contains this the discrepancy between the embedding similarities and graph similarities

File: docs/use_cases/knowledge_graph_embedding.md [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+
+The simplest approach is to consider nodes that are connected by an edge as similar. Then, the task of learning node embeddings can be defined as a classification task. Given the embeddings of two nodes and a relation, the task is to determine how likely it is that they are similar (connected).
```
Comment (morkapronczay) [inline docs/use_cases/knowledge_graph_embedding.md:37]: I don't get this.Isn't it the thing that connected nodes' embeddings should be similar to each other?

File: docs/use_cases/knowledge_graph_embedding.md [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+| HitRate@10 |  0.010 | 0.0436 | **0.307** |
+
+The table above clearly shows that LLM performs three times better than the method of randomly ordering nodes. It is also clear that KGE significantly outperforms LLM, with hit rates almost ten times higher. Notably, DistMult can find the correct answer on its first attempt more frequently than LLM can in ten attempts. It's important to point out that we used only 250-dimensional embeddings with DistMult, while LLM outputs 768-dimensional vectors, which makes the performance difference even more noticeable.
+
+These results strongly support that KGE is more suitable than LLMs for tasks where relational information is super important.
+
+## Limitations
+
+One reason why the LLM approach might struggle with performance is due to the formulation used, where each node is mapped to a sequence of sentences describing its connections. This method tends to overload the input text with an extensive amount of information for a single node. LLMs are typically not trained to handle such broad and diverse information within a single context; their strength lies in processing more focused and specific textual information.
```
Comment (morkapronczay) [inline docs/use_cases/knowledge_graph_embedding.md:217]: I think you should rephrase this. It's not that we cheated, it's that relational data is hard to be described to and grasped by an LLM. That is an inherent weakness of the approach, hence we should use KGEs.

**Review by morkapronczay — CHANGES_REQUESTED** [review]
All in all great work, loved the presentation of the results and the conclusions drawn!
Feel free to dispute any of my comments, if needed we have a quick call too.

File: docs/use_cases/knowledge_graph_embedding.md (hunk 6 above)
Comment (ricsi98) [reply docs/use_cases/knowledge_graph_embedding.md:37]: It means exactly that. By using dot product as the similarity function, and cross entropy on the presence of an edge as the loss, that will force connected nodes to have similar embeddings (high dot product value) while not connected pairs to have low similarity (low dot product).

File: docs/use_cases/knowledge_graph_embedding.md (hunk 3 above)
Comment (morkapronczay) [reply docs/use_cases/knowledge_graph_embedding.md:7]: I see your approval, but you didn't change the text here, right? Is it in progress, or do you think it should stay like this?

File: docs/use_cases/knowledge_graph_embedding.md (hunk 6 above)
Comment (morkapronczay) [reply docs/use_cases/knowledge_graph_embedding.md:31]: I get it now :) I think you should extend this paragraph to include this, too.

File: docs/use_cases/knowledge_graph_embedding.md (hunk 3 above)
Comment (ricsi98) [reply docs/use_cases/knowledge_graph_embedding.md:7]: I included that in the previous paragraph. Do you want me to emphasize this by including it here too?

Comment (morkapronczay) [reply docs/use_cases/knowledge_graph_embedding.md:7]: yes, sorry indeed, I should have looked at the diff not just the paragraph I commented on. Good!

**Review by morkapronczay — APPROVED** [review]
